rviz_common/include/rviz_common/message_filter_display.hpp

```diff
@@ -103,6 +103,27 @@ class MessageFilterDisplay : public _RosTopicDisplay
 protected:
   void updateTopic() override
   {
+    // Change the QoS according to the info in the ROS Graph
+    const auto & publishers = rviz_ros_node_.lock()->get_raw_node()->get_publishers_info_by_topic(
+      topic_property_->getTopicStd());
+
+    if (!publishers.empty()) {
+      // Check if there is any Subscriber with the same QoS as set
+      bool any_publisher_qos_compatible = false;
+      for (const auto & publisher : publishers) {
+        if (qos_profile == publisher.qos_profile()) {
+          any_publisher_qos_compatible = true;
+          break;
+        }
+      }
+
+      // If not found, select one valid
+      if (!any_publisher_qos_compatible) {
+        qos_profile = publishers[0].qos_profile();
+        qos_profile_property_->setQoSProfile(qos_profile);
+      }
+    }
```

Comment:
Logic seems fine: If there is at least one publisher with the same QoS settings, keep current settings. If not, then take the QoS of the first publisher.

It might be handy to warn about having multiple publishers with different QoS, if that happens. But not sure if that is RViz's role.

Reply:
Let's see what RViz2 maintainers say about this